### PR TITLE
feat: Kpt CLI Function discovery

### PR DESCRIPTION
**Pull request by yuwenma**
### Description

This PR extends the `kpt fn eval` autocompletion to help user to discover functions by using the porch function fields, which store the OCI image annotations. the work on porch side has been merged  https://github.com/GoogleContainerTools/kpt/pull/2972 

- Define `NewPorchFunctionClient`, which provides the restclient to query porch orchestrator. Set the restulapi call timeout  to be 3 seconds. This timeout is used to determine whether porch is enabled on the cluster side.  
- `ListKeywords` iterates the porch functions' `keywords` field, and suggest users a unique keyword list.
- `ListFunctions` iterates the porch functions and filter functions with the matching function type (default to `mutator`) and the `keywords` (if not given, accept all) a well as  the existing [catalog.kpt.dev](https://catalog.kpt.dev/) functions  
- All suggested functions will be filtered based on the function type (`mutator` or `validator`)/
 
### New UX experience 
- Add `--type` flag to register the autocomplete with value `mutator`, `validator`, and store the evaluated functions to the right kptfile list: `.pipeline.mutators` if mutator type, `.pipeline.validators` if validator type. 
- Add `--keywords` flag which pull the porch functions to get the `function.spec.keywords` and provide users a unique keywords list
- Modify the `--image` autocompletion function to rely on NewPorchFunctionClient to
  -  decide  whether porch is used
  -  autocomplete showing the `keywords` from porch functions
  - autocomplete showing the `functions` from porch and catalog.kpt.dev, filter by function type and keywords. 

### Fixes
- Change the image annotation prefix from `kpt.dev.fn` to `dev.kpt.fn` to follow the reverse domain convention
- Fix the annotation mapping for `description` and `documentationURL`

### Update:
- `--type` related code is merged in https://github.com/GoogleContainerTools/kpt/pull/2975

### TODO:
- Update `AddDefaultImagePathPrefix`  to append porch function prefix.  
- Redesign the kpt-functions-catalog release workflow, add the function meta resource to function OCI image annotation before pushing to [gcr.io/kpt-fn](gcr.io/kpt-fn), including the meta data in [https://catalog.kpt.dev/catalog-v2.json](https://catalog.kpt.dev/catalog-v2.json)
- Modify the existing autocompletion to merge the logic of fetching functions from catalog.kpt.fn with the porch function filtering. Use the same rules (by function type, by keywords) to suggest functions and keywords in `kpt fn eval` autocompletion.

### CODE REVIEW

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go [hunk 1]
```diff
@@ -164,43 +175,58 @@ func (r *EvalFnRunner) NewFunction() *kptfile.Function {
...
+// SaveFnToKptfile adds the evaluated function and its arguments to Kptfile `pipeline.mutators` list.
+func (r *EvalFnRunner) SaveFnToKptfile() {
+	pr := printer.FromContextOrDie(r.Ctx)
+	kf, err := pkg.ReadKptfile(filesys.FileSystemOrOnDisk{}, r.RunFns.Path)
+	if err != nil {
+		pr.Printf("function not added: Kptfile not exists\n")
+		return
+	}
+	if kf.Pipeline == nil {
+		kf.Pipeline = &kptfile.Pipeline{}
```
Comment (martinmaly) [inline thirdparty/cmdconfig/commands/cmdeval/cmdeval.go:214]: this may have side effect on the Kptfile if no function is added; this will still add the Pipeline section, albeit empty one.

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go [hunk 2]
```diff
@@ -164,43 +175,58 @@ func (r *EvalFnRunner) NewFunction() *kptfile.Function {
...
+	}
+	switch r.FnType {
+	case "mutator":
+		if kf.Pipeline.Mutators == nil {
+			kf.Pipeline.Mutators = []kptfile.Function{}
+		}
+		newFns := r.updateFnList(kf.Pipeline.Mutators)
+		kf.Pipeline.Mutators = newFns
```
Comment (martinmaly) [inline thirdparty/cmdconfig/commands/cmdeval/cmdeval.go:222]: append works on nil lists to all this can be simplified as:

kf.Pipeline.Mutators = r.updateFnList(kf.Pipeline.Mutators)

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go [hunk 3]
```diff
@@ -41,14 +41,23 @@ func GetEvalFnRunner(ctx context.Context, parent string) *EvalFnRunner {
...
 	r.Command.Flags().StringVarP(&r.Dest, "output", "o", "",
 		fmt.Sprintf("output resources are written to provided location. Allowed values: %s|%s|<OUT_DIR_PATH>", cmdutil.Stdout, cmdutil.Unwrap))
 	r.Command.Flags().StringVarP(
 		&r.Image, "image", "i", "", "run this image as a function")
 	_ = r.Command.RegisterFlagCompletionFunc("image", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
-		return cmdutil.FetchFunctionImages(), cobra.ShellCompDirectiveDefault
+		return cmdutil.SuggestFunctions(cmd), cobra.ShellCompDirectiveDefault
+	})
+	r.Command.Flags().StringArrayVarP(
+		&r.Keywords, "keywords", "k", []string{}, "discover the keywords correlated functions")
```
Comment (martinmaly) [inline thirdparty/cmdconfig/commands/cmdeval/cmdeval.go:53]: I'm having hard time figuring out the help text. I am also having even harder time coming up with a better suggestion :)

my best try: "filter functions that match one or more keywords"

File: porch/repository/pkg/oci/oci.go [hunk 4]
```diff
@@ -215,7 +215,7 @@ func GetSliceFromAnnotation(key string, manifest *v1.Manifest) []string {
 }
 
 func GetSingleFromAnnotation(key string, manifest *v1.Manifest) string {
-	if val, ok := manifest.Annotations[FunctionTypesKey]; ok {
+	if val, ok := manifest.Annotations[key]; ok {
```
Comment (martinmaly) [inline porch/repository/pkg/oci/oci.go:218]: good catch!

File: internal/porchutil/function.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		if !ContainsKeywords(function, keywords) {
+			continue
+		}
+		suggestedFunctions = append(suggestedFunctions, GetFunctionShortName(function))
+	}
+	return suggestedFunctions, nil
+}
+
+func visitFunctions() ([]v1alpha1.Function, error) {
+	kubeconfig, ok := os.LookupEnv("KUBECONFIG")
```
Comment (martinmaly) [inline internal/porchutil/function.go:131]: hmm ... I've never needed to do this; the k8s go-client has wrappers that take care of it.

This is what I am doing (not claiming that it is correct though :) but it _seems_ to work:

https://github.com/GoogleContainerTools/kpt/blob/main/internal/util/porch/client.go#L30

File: internal/porchutil/function.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var functions v1alpha1.FunctionList
+	raw, err := clientset.RESTClient().Get().Timeout(PorchExpiration).AbsPath(functionAPI).DoRaw(context.TODO())
```
Comment (martinmaly) [inline internal/porchutil/function.go:146]: interesting ... this will list all functions in all namespaces in the k8s cluster.
not something to figure out now but the namespace-deness of k8s is making things less straightforward since it is unclear which namespace we'd be querying if we had to pick.

So this is fine for now.

File: internal/porchutil/function.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("fail to call %v:%v", functionAPI, err)
+	}
+	err = json.Unmarshal(raw, &functions)
```
Comment (martinmaly) [inline internal/porchutil/function.go:150]: this shouldn't be needed either. I find the controller-runtime client does the unmarshaling too.

File: internal/porchutil/function.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	var keywords []string
+	for keyword := range keywordsMap {
+		keywords = append(keywords, keyword)
+	}
+	return keywords, nil
+}
+
+func getKeywordsFromFlag(cmd *cobra.Command) []string {
+	flagVal := cmd.Flag("keywords").Value.String()
```
Comment (martinmaly) [inline internal/porchutil/function.go:93]: Cobra doesn't have facility to accept lists as flag values? I'm just wondering if we can avoid the custom parser.

File: internal/porchutil/function.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+			if actual != expected {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// ListKeywords returns all the unique keywords from Porch Functions.
+func (p *PorchFunctionClient) ListKeywords(cmd *cobra.Command) ([]string, error) {
```
Comment (martinmaly) [inline internal/porchutil/function.go:74]: another way to do this could be that this function accepts:

`func match(*Function) bool`

and we can compose the matching on type and on keywords using that.

To me the 'red' flag is the interdependence of Porch Client (which conceptually can/should be purely around Porch API concepts) and cobra Commands, resulting in leaky abstraction

File: internal/porchutil/function.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+var (
+	functionAPI     = "/apis/" + v1alpha1.SchemeGroupVersion.String() + "/functions"
+	PorchExpiration = time.Duration(expirationNanoSecond) * time.Second
+)
+
+// NewPorchFunctionClient provides the PorchFunctionClient which connects to Porch server
+// and gets the functions.
+func NewPorchFunctionClient() *PorchFunctionClient {
+	functions, err := visitFunctions()
```
Comment (martinmaly) [inline internal/porchutil/function.go:32]: nit: visitFunctions signals (to me) that the function is iterating over an existing data, rather than fetching it. consider whether another name would convey the purpose more clearly

File: internal/porchutil/function.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
+package porchutil
```
Comment (martinmaly) [inline internal/porchutil/function.go:1]: missing header

also, we have this package: https://github.com/GoogleContainerTools/kpt/tree/main/internal/util/porch

would it be a good fit for this kind of utility? (please advise, I am not super familiar with kpt repo organization conventions)

File: internal/porchutil/function.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/kubernetes"
+
+	// Enable the GCP Authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const expirationNanoSecond = 3
```
Comment (martinmaly) [inline internal/porchutil/function.go:22]: that's very short amount of time :)

File: internal/porchutil/function.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+	if err != nil {
+		return &PorchFunctionClient{IsPorchAvailable: false, Functions: nil}
+	}
+	return &PorchFunctionClient{IsPorchAvailable: true, Functions: functions}
+}
+
+type PorchFunctionClient struct {
+	IsPorchAvailable bool
+	Functions        []v1alpha1.Function
+}
```
Comment (martinmaly) [inline internal/porchutil/function.go:42]: slight conceptual misalignment. this struct is actually not a client. I wonder if it would work as well to have a simpler interface, possibly avoiding the IsPorchAvailable bool and return list of functions directly?

(I am not feeling too strongly though)

**Review by martinmaly — APPROVED** [review]
some comments for your consideration.

File: e2e/testdata/fn-eval/save-fn/validator-type/.expected/diff.patch [hunk 14]
```diff
@@ -0,0 +1,34 @@
+diff --git a/Kptfile b/Kptfile
+index d9e2f05..c6e80ef 100644
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply e2e/testdata/fn-eval/save-fn/validator-type/.expected/diff.patch:2]: they are stable.

File: e2e/testdata/fn-eval/save-fn/validator-type/.expected/config.yaml [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+testType: eval
+image: set-namespace:v0.1.3
+args:
+  namespace: staging
+stdErr: |
+  [RUNNING] "gcr.io/kpt-fn/set-namespace:v0.1.3"
+  [PASS] "gcr.io/kpt-fn/set-namespace:v0.1.3" in 0s
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply e2e/testdata/fn-eval/save-fn/validator-type/.expected/config.yaml:21]: Exactly.

File: internal/porchutil/function.go (hunk 12 above)
Comment (yuwenma) [reply internal/porchutil/function.go:22]: The IETF says [current request timeout implementation select times between 30-120s](https://tools.ietf.org/id/draft-thomson-hybi-http-timeout-00.html#:~:text=Knowing%20the%20time%20available%20to,empirically%20determined%20to%20be%20safe.)

Maybe we can set that to 30s?

Comment (martinmaly) [reply internal/porchutil/function.go:22]: this is for tab completion, correct? IMO 30s is too long.
let's try some value now (I'd recommend using the Duration type) and we'll adjust.

File: internal/porchutil/function.go (hunk 13 above)
Comment (yuwenma) [reply internal/porchutil/function.go:42]: that's true. I defined the struct to make the caller code be more specific that only when porch available will we use  porch functions
```
if client.IsPorchAvailable():
   client.ListFunctions
```
I don't have a strong preference but agree the naming is misaligned. I'll update.

File: internal/porchutil/function.go (hunk 11 above)
Comment (droot) [reply internal/porchutil/function.go:1]: I think we should use one `porch` client for all porch APIs. This seems to duplicate the code in `internal/util/porch` ?

File: internal/porchutil/function.go (hunk 12 above)
Comment (droot) [reply internal/porchutil/function.go:22]: I wonder what the experience is like in `porch` disconnected mode, does it fail fast ? (you might need two timeout, one for connectionTimeout, another one timing out on waiting from response).

File: internal/porchutil/function.go (hunk 7 above)
Comment (droot) [reply internal/porchutil/function.go:150]: @yuwenma would recommend using the existing porch client, it uses `controller-runtime` underneath and that does all of this for you for free.

File: internal/util/cmdutil/cmdutil.go [hunk 16]
```diff
@@ -178,6 +179,44 @@ func CheckDirectoryNotPresent(outDir string) error {
...
+	functions = append(functions, FetchFunctionImages()...)
+	var dedup []string
+	fnMap := map[string]bool{}
+	for _, function := range functions {
+		if _, ok := fnMap[function]; !ok {
+			dedup = append(dedup, function)
+			fnMap[function] = true
+		}
+	}
+	return dedup
```
Comment (droot) [inline internal/util/cmdutil/cmdutil.go:204]: IMO, we should fire off both lookups in parallel (porch and functionImages) so that autocomplete UX feels instantaneous (more important from `porch disconnected` mode)

File: go.sum [hunk 17]
```diff
@@ -74,6 +74,10 @@ github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym
 github.com/DataDog/datadog-go v3.2.0+incompatible/go.mod h1:LButxg5PwREeZtORoXG3tL4fMGNddJ+vMq1mwgfaqoQ=
 github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220404223850-4fe7ba260b65 h1:HTV58N6lLjmq4XNhlxzU/u435HO8zzXIYQET6WJc8aI=
 github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220404223850-4fe7ba260b65/go.mod h1:cMUBP0fY1F/1noOO1jJoIHM3R8Cp1BYWIG9XAZET5gI=
+github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220406191334-921abb821a86 h1:CczJ3rznKSvyj9VDpLZdTcRy2ihGdeHR7ImvlufBwkQ=
+github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220406191334-921abb821a86/go.mod h1:51Vk7QZ+XUzHCvQBi7t9tiWqTXvy6T13cv/inUXJJ0s=
+github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220407000100-7a5545349d71 h1:ciqpo48GR1x4CI46l2t0+J1WXTQj0wSKI8ckixStVcQ=
+github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220407000100-7a5545349d71/go.mod h1:51Vk7QZ+XUzHCvQBi7t9tiWqTXvy6T13cv/inUXJJ0s=
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply go.sum:80]: wow, how did you catch that... TBH, I never really pay attention to go.mod

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go [hunk 18]
```diff
@@ -164,43 +175,58 @@ func (r *EvalFnRunner) NewFunction() *kptfile.Function {
...
 	if !found {
-		newMutators = append(newMutators, *r.NewFunction())
+		newFns = append(newFns, *r.NewFunction())
 		pr.Printf("adding function to Kptfile\n")
 	}
-	kf.Pipeline.Mutators = newMutators
+	return newFns
+}
+
+// SaveFnToKptfile adds the evaluated function and its arguments to Kptfile `pipeline.mutators` list.
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply thirdparty/cmdconfig/commands/cmdeval/cmdeval.go:205]: good catch. thanks

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go (hunk 1 above)
Comment (yuwenma) [reply thirdparty/cmdconfig/commands/cmdeval/cmdeval.go:214]: This is called after the function run, so we always expect new functions to be added (unless it already exists in Kptfile).

File: internal/porchutil/function.go (hunk 10 above)
Comment (yuwenma) [reply internal/porchutil/function.go:32]: Sounds good. Yeah, I actually did some code refactoring before uploading but didn't rethink about the function names.

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go (hunk 1 above)
Comment (martinmaly) [reply thirdparty/cmdconfig/commands/cmdeval/cmdeval.go:214]: ack.  thanks

File: internal/porchutil/function.go (hunk 8 above)
Comment (yuwenma) [reply internal/porchutil/function.go]: Yeah, I explored this a little bit. I used `StringArrayVarP` for "keywords", which users can provide an array of keywords in the form as `--keywords XX --keywords YY`. But when parsing the flag from commands, it does not provide such options.

File: internal/porchutil/function.go (hunk 9 above)
Comment (yuwenma) [reply internal/porchutil/function.go]: really like this suggestion! thank you. Refactored the code.

File: internal/util/cmdutil/cmdutil.go (hunk 16 above)
Comment (yuwenma) [reply internal/util/cmdutil/cmdutil.go:203]: I'll do that in the follow up PR since I also plan to make some major changes to the `fetchFunctions` (including OCI annotations and filter functions by annotation)

File: internal/porchutil/function.go (hunk 11 above)
Comment (yuwenma) [reply internal/porchutil/function.go]: Yeah, I didn't noticed that client code. Updated.

File: internal/porchutil/function.go (hunk 13 above)
Comment (yuwenma) [reply internal/porchutil/function.go]: set timeout to 10.

File: internal/porchutil/function.go (hunk 6 above)
Comment (martinmaly) [reply internal/porchutil/function.go]: agreed. let's keep this; for autocomplete being broader is good, and we can adjust in the future.

File: thirdparty/cmdconfig/commands/cmdeval/cmdeval.go (hunk 3 above)
Comment (martinmaly) [reply thirdparty/cmdconfig/commands/cmdeval/cmdeval.go]: quite possibly. IMO we need to (find a way) to communicate that the keywords here are used to filter functions, using the keywords stored on the functions.

(hopefully better try than my previous one): "Filter function; include functions whose keywords match one or more of the keywords provide in this flag"

(may be too wordy though; perhaps we can trim it down)

File: internal/porchutil/function.go (hunk 8 above)
Comment (martinmaly) [reply internal/porchutil/function.go]: I see. re-reading the code it seems that for the completion, less arg parsing has been done. That _may_ mean that if customer enters --keywords XX --keywords YY ... the code may not see both options?

File: internal/util/porch/function.go [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+
+	// Enable the GCP Authentication plugin
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const expirationNanoSecond = 10
+
+var (
+	functionAPI     = "/apis/" + v1alpha1.SchemeGroupVersion.String() + "/functions"
```
Comment (martinmaly) [inline internal/util/porch/function.go:35]: GitHub observers this is unused

File: internal/util/porch/function.go [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+	flagVal := cmd.Flag("keywords").Value.String()
+	flagVal = strings.TrimPrefix(flagVal, "[")
+	flagVal = strings.TrimSuffix(flagVal, "]")
+	splitted := strings.Split(flagVal, ",")
+	var trimmed []string
+	for _, val := range splitted {
+		trimmed = append(trimmed, strings.TrimSpace(val))
+	}
+	return trimmed
+}
```
Comment (martinmaly) [inline internal/util/porch/function.go:160]: [nit] I'd recommend moving this function out to the command-related code, i.e. into cmdutil.go to keep the porch code unaware of cobra.

**Review by martinmaly — APPROVED** [review]
Couple more nits but LGTM once tests pass.

File: go.mod [hunk 21]
```diff
@@ -3,7 +3,7 @@ module github.com/GoogleContainerTools/kpt
 go 1.17
 
 require (
-	github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220404223850-4fe7ba260b65
+	github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220407000100-7a5545349d71
```
Comment (martinmaly) [inline go.mod:6]: we now need `github.com/GoogleContainerTools/kpt/porch/api v0.0.0-20220404223850-4fe7ba260b65` which is already at HEAD so this file can be reverted.

Comment (yuwenma) [reply go.mod:6]: thanks.

Comment (martinmaly) [reply go.mod]: I'm sorry ... I was wrong. I thought my PR was already merged; alas, it wasn't. your code was correct.

Comment (yuwenma) [reply go.mod]: I pulled the latest and got v0.0.0-20220411164219-e3555a1d90a9. Shall I use this instead?
